### PR TITLE
Profiling

### PR DESCRIPTION
**Pull request by spall**
Write information about re-running commands to a file.  Open to suggestions on more readable/better ways to display this information.

### CODE REVIEW

File: src/Development/Rattle/Server.hs [hunk 1]
```diff
@@ -230,6 +255,29 @@ cmdRattleRun rattle@Rattle{..} cmd@(Cmd opts args) startTimestamp hist msgs = do
                     let f hasher xs = mapMaybeM (\x -> fmap (\(mt,h) -> (x,mt,h)) <$> hasher x) $ filter (not . skip) xs
                     touch <- Touch <$> f hashFileForward (tRead touch) <*> f hashFile (tWrite touch)
                     touch <- generateHashForwards cmd [x | HashNonDeterministic xs <- opts2, x <- xs] touch
+                    when (isJust $ rattleDebug options) $
```
Comment (ndmitchell) [inline src/Development/Rattle/Server.hs]: Better to use whenJust, which gives you the thing inside the just, and let syou avoid the fromJust

File: src/Development/Rattle/Server.hs [hunk 2]
```diff
@@ -230,6 +255,29 @@ cmdRattleRun rattle@Rattle{..} cmd@(Cmd opts args) startTimestamp hist msgs = do
...
+                      -- hashtable of files -> hashes from first trace in hist
+                      let ht = foldl (\x (fp, _, h) -> Map.insert fp h x)
+                               Map.empty $ if null hist
+                                           then []
+                                           else tWrite $ tTouch $ head hist in
```
Comment (ndmitchell) [inline src/Development/Rattle/Server.hs]: Would this be clearer as:

```
case hist of
   [] -> Map.empty
   [maybe a view pattern -> x] -> Map.fromList ...
```

File: src/Development/Rattle/Server.hs (hunk 1 above)
Comment (ndmitchell) [inline src/Development/Rattle/Server.hs]: Could this entire block be extracted to a separate function?

File: src/Development/Rattle/Server.hs [hunk 3]
```diff
@@ -194,12 +203,28 @@ cmdRattleStarted rattle@Rattle{..} cmd s msgs = do
...
+helper :: Monad m => (a -> m (Maybe b)) -> [a] -> m ([a], [b])
+helper _ [] = return ([], [])
+helper f (x:xs) = do
+  res <- f x
+  (as,bs) <- helper f xs
+  case res of
+    Nothing -> return (x:as, bs)
+    Just y -> return (as, y:bs)
+
+g :: (Hashable a, Eq a, Monad m) => (a -> m Bool) -> [a] -> m (Maybe (Set.HashSet a))
```
Comment (ndmitchell) [inline src/Development/Rattle/Server.hs]: Is this a filterM with a conditional on the outside, so [] => Nothing, xs => Just . Set.fromList?

**Comment by ndmitchell** [comment]
Is this something we should close and tackle as a separate PR?

**Comment by spall** [comment]
That is fine.

**Pull request closed without merging**